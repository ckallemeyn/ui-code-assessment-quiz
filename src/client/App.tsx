import * as React from 'react';
import { useState, useEffect } from 'react';
import { unescapeStr } from '../utils/questionUtils'

interface Question {
  category: string,
  type: string,
  difficulty: string,
  question: string,
  correct_answer: string,
  incorrect_answers: string[]
}

interface IState {
  questions: Question[],
  currentQuestion: Question | any,
  correctAnswers: number,
  incorrectAnswers: number,
  questionsAnswered: number,
  finalScorePercentage: number,
}


/*
initialState = {
  questions: [],
  currentQuestion: {},
  correctAnswers: 0,
  incorrectAnswers: 0,
  questionsAnswered: 0,
  finalScorePercentage: 0,
}
*/


export const App = () => {
  const [questions, setQuestions] = useState<Question[] | []>([]);
  const [currentQuestion, setCurrentQuestion] = useState<Question | any>({});

  useEffect(() => {
    fetch('http://localhost:4000/api/questions')
      .then(res => res.json())
      .then(q => {
        const data: any = q.results;
        const firstQuestion: any = data[0];
        console.log(unescapeStr(firstQuestion.question));
        setQuestions(data);
        setCurrentQuestion(firstQuestion);
      })
      .catch(error => console.error('unable to retrieve data at this time please try again', error))
  }, [])



  return (
    <>
      {(questions && currentQuestion) && questions.length < 1
        ? <div>Loading...</div>
        : <div style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}>
          <h1>{unescapeStr(currentQuestion.question)}</h1>
            <button>Next</button>
          </div>
      }
    </>
  );
}
