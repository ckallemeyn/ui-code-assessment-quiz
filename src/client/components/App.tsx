import * as React from "react";
import { useEffect, useReducer } from "react";
import { Container, Menu, MenuItem } from 'semantic-ui-react'
import { configureAnswers, handleQuestionChange, shuffle } from "../../utils/questionUtils";
import QuestionForm from "./QuestionForm";
import QuestionContext from "../contexts/QuestionContext";
import mainReducer from '../reducers/index'

interface Question {
  category: string;
  type: string;
  difficulty: string;
  question: string;
  correct_answer: string;
  incorrect_answers: string[];
}

interface IState {
  questions: Question[];
  currentQuestion: Question | any;
  idx: number;
  answers: [] | any;
  currentAnswer: string | any;
  correctAnswers: number;
  incorrectAnswers: number;
  questionsAnswered: number;
  finalScorePercentage: number;
  isSummaryVisible: boolean;
  isWarningMessageVisible: boolean;
}

const initialState: IState = {
  questions: [],
  currentQuestion: {},
  idx: 0,
  answers: [],
  currentAnswer: "",
  correctAnswers: 0,
  incorrectAnswers: 0,
  questionsAnswered: 0,
  finalScorePercentage: 0,
  isSummaryVisible: false,
  isWarningMessageVisible: false,
};


export const App = () => {
  const [state, dispatch] = useReducer(mainReducer, initialState);
  const { currentAnswer } = state;

  const handleChange = (_e: any, { value }: any) => {
    const payload = {
      currentAnswer: value,
    };

    return dispatch({ type: "ONCHANGE", payload });
  };

  const updateQuestion = (e: any) => handleQuestionChange(e, state, dispatch)


  useEffect(() => {
    fetch("http://localhost:4000/api/questions")
      .then((res) => res.json())
      .then(({ results }: any) => {
        const questions = shuffle(results);
        const currentQuestion = questions[0];
        const payload: any = {
          answers: configureAnswers(currentQuestion),
          questions,
          currentQuestion,
        };

        return dispatch({ type: "SET_DATA", payload });
      })
      .catch((error) => console.error("Unable to retrieve data, please try again", error))
  }, []);

  return (
    <>
      <QuestionContext.Provider value={{ state, currentAnswer, handleChange, dispatch }}>
        <Menu fixed='top' inverted>
          <MenuItem as='a' header>
            QUIZR
          </MenuItem>
        </Menu>
        <Container style={{ marginTop: '4em' }}>
          {state.questions && state.questions.length < 1 ? (
            <div style={{ display: "flex", flexDirection: "column", alignItems: "center",}}>
              Loading...
            </div>
          ) : (
            <div>
              <QuestionForm state={state} updateQuestion={updateQuestion} />
            </div>
          )}
        </Container>
      </QuestionContext.Provider>
    </>
  );
};
