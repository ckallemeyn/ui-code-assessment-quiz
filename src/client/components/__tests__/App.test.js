import React from "react";
import { render } from "@testing-library/react";
import { App } from "../App";

describe('Main App Component', () => {
  test('it should render', () => {
    const { container } = render(<App />)

    expect(container).toMatchSnapshot();
  })
})
