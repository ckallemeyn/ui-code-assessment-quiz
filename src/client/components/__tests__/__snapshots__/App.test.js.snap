// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main App Component it should render 1`] = `
<div>
  <div
    class="ui inverted top fixed menu"
  >
    <a
      class="header item"
    >
      QUIZR
    </a>
  </div>
  <div
    class="ui container"
    style="margin-top: 4em;"
  >
    <div
      style="display: flex; flex-direction: column; align-items: center;"
    >
      Loading...
    </div>
  </div>
</div>
`;
